### ISSUE TITLE
Feature request: on INSERT/UPDATE trigger events

### ISSUE THREAD

**Issue by pslavov** [2ndQuadrant/bdr#83]
Hi,
is there some way to trigger an event function on insert/update change on the replicated tables?
I want to do some masking on the replicated data, and I know that masking a specific replicated column is too much to ask :)

**Comment by anarazel** [comment]
> is there some way to trigger an event function on insert/update change
> on the replicated tables?

No, at the moment not. It'd require adding a bit more overhead - using a
full executor startup + using nodeModifyTable.c - or reimplementing
trigger invocations. The latter will also imply some overhead.

I know that Petr/@PJMODOS was interested in the topic though.

**Comment by ringerc** [comment]
We've had quite a number of people interested in this - being able to run a trigger or equivalent on the row at apply time to modify the row, skip it, and/or perform actions with side effects when the row is applied.

A few people have wanted to send `NOTIFY` messages. Some have wanted to transform the row into JSON and send it to an external message queue. All sorts of things, some of which might make sense under BDR/UDR, and others that probably make more sense as a separate output plugin.

As Andres notes, it'd mean a heavier-weight apply. Right now we modify the heap at quite a low level and invoke index maintenance directly.

**Comment by anarazel** [comment]
> We've had quite a number of people interested in this - being able to
> run a trigger or equivalent on the row at apply time to modify the
> row, skip it, and/or perform actions with side effects when the row is
> applied.

FWIW, I think it'd be a mistake to implement this at this time. This has
the potential of making things vastly more complicated and failure
prone. And we're not exactly minimal on either today. This seems like a
2.0 time feature.

> A few people have wanted to send `NOTIFY` messages. Some have wanted
> to transform the row into JSON and send it to an external message
> queue. All sorts of things, some of which might make sense under
> BDR/UDR, and others that probably make more sense as a separate output
> plugin.

Neither of the two you named here seem to make sense to do via UDR/BDR.

With regards to the message queue, bottled water might be interesting:
https://github.com/confluentinc/bottledwater-pg
http://blog.confluent.io/2015/04/23/bottled-water-real-time-integration-of-postgresql-and-kafka/

**Comment by ringerc** [comment]
I entirely agree, it's 2.0 material, and only for some use cases even then.
